### DISCUSSION TITLE
Running on Cloudflare Workers?

### DISCUSSION THREAD

**Discussion by faldonis** [ConsenSys-archive/ganache#4418] [Q&A] [answered]
Does any know about the feasibility of this? Ganache runs in the browser so it appears to be technically possible.  Is there anything preventing Ganache from running in the service worker-like context of Cloudflare Workers?

**Comment by davidmurdoch** [comment, marked as answer]
I've done it, however the cold starts were a problem, as workers must load the ganache code from scratch every time, and this _sometimes_ exceeded the amount of time workers are given to start up.

From [Cloudflare docs](https://developers.cloudflare.com/workers/platform/limits/#worker-startup-time):

> A Worker must be able to be parsed and execute its global scope within 200 ms. Script size can impact startup because there’s more code to parse and evaluate.

But if you lazy load ganache (inside the worker handler) it still has problems because of [CPU time limits](https://developers.cloudflare.com/workers/platform/limits/#cpu-runtime):

> A Worker may consume up to 10 milliseconds on the Free plan.

If Ganache can't be parsed and evaluate by Workers within 200ms we certainly won't be able to do it in 10ms!

So this means you'd need use the paid plan just because ganache is so big right now.

---

I think the issue is that we ship all of Ganache in a single file: https://cdn.jsdelivr.net/npm/ganache@7.8.0/dist/web/ganache.min.js. If instead, we adeduplicated some of the dependencies that end up in the bundle (I think bn.js gets in there a bunch of times) and/or split it into smaller parts that could each be lazy-loaded when needed, it could work.

**Reply by faldonis** [reply]
Thanks for the detailed reply @davidmurdoch . I've seen duplicated bn.js's increase bundle size by 40% in a previous project.

Did you have to implement a storage layer in the Worker to make it work? Got any code to share to save me some cycles?

**Reply by davidmurdoch** [reply]
Yeah, I thought I had bn.js deduped a long time ago, but it seems to have snuck back in at some point.

I was just experimenting and didn't get far enough to try to add persistence. There is a `database.db` option you can use, it requires an abstract-leveldown interface. We automatically use `browser-level` (backed by indexeddb) in the browser for storage, but you would probably want to add your own layer based on Cloudlare KV or maybe you could use a Cloudflare Durable. If you don't care about persistence memdown works well.

**Reply by faldonis** [reply]
I started looking into it and wanted to check a premise before diving in further. Does ganache need to be running locally to intercept RPC traffic from a dapp or can it be running on a remote machine?

**Reply by davidmurdoch** [reply]
It can be run remote, but not in browser (since we can't bind to the host/port from the browser).

In Node its perfectly fine, however, it doesn't currently allow for configuring SSL, so you'll want to have something else (like Cloudflare SSL, ngnix, etc) take care of SSL termination.
